redbot/core/core_commands.py

```diff
@@ -454,11 +464,89 @@ def get_bot_uptime(self, *, brief: bool = False):
             await ctx.send("No exception has occurred yet")
 
     @commands.command()
-    @checks.is_owner()
-    async def invite(self, ctx: commands.Context):
+    @commands.check(CoreLogic._can_get_invite_url)
+    async def invite(self, ctx):
         """Show's Red's invite url"""
         await ctx.author.send(await self._invite_url())
 
+    @commands.group()
+    @checks.is_owner()
+    async def inviteset(self, ctx):
+        """Setup the bot's invite"""
+        if not ctx.invoked_subcommand:
+            await ctx.send_help()
+
+    @inviteset.command()
+    async def public(self, ctx, confirm: bool = False):
+        """
+        Define if the command should be accessible\
+        for the average users.
+        """
+        if not self.bot.user.bot:
+            await ctx.send("I'm not a bot account. I have no invite URL.")
+            return
+        if await self.bot.db.invite_public():
+            await self.bot.db.invite_public.set(False)
+            await ctx.send("The invite is now private.")
+            return
+        app_info = await self.bot.application_info()
+        if not app_info.bot_public:
+            await ctx.send(
+                "I am not a public bot. That means that nobody except "
+                "you can invite me on new servers.\n\n"
+                "You can change this by ticking `Public bot` in "
+                "your token settings: "
+                "https://discordapp.com/developers/applications/me/{0}".format(self.bot.user.id)
+            )
+            return
+        if not confirm:
+            await ctx.send(
+                "You're about to make the `{0}invite` command public. "
+                "All users will be able to invite me on their server.\n\n"
+                "If you agree, you can type `{0}inviteset public yes`.".format(ctx.prefix)
+            )
+        else:
+            await self.bot.db.invite_public.set(True)
+            await ctx.send("The invite command is now public.")
+
+    @inviteset.command()
+    async def perms(self, ctx, level: int):
+        """
+        Make the bot create its own role with permissions on join.
+
+        The bot will create its own role with the desired permissions\
+        when he join a new server. This is a special role that can't be\
+        deleted or removed from the bot.
+
+        For that, you need to give a valid permissions level.
+        You can generate one [here](https://discordapi.com/permissions.html).
+
+        Please note that you might need the two factor authentification for\
+        some permissions.
+        """
+        await self.bot.db.invite_perm.set(level)
+        await ctx.send("The new permissions level has been set.")
+
+    @inviteset.command()
```

Comment:
OAuth Redirects should not be user facing in the settings here in my opinion, as Red itself does nothing to support this, and this requires further settings properly configured on the developer panel to function.

Any developer who needs this, should have their code remove the core invite command and replace with their own. If this is beyond their capability, this setting is useless to them anyway.

Reply:
This is something that can be useful for users and telling them to edit the bot is not a solution (especially with how hard it is to edit pip packages and keep them up-to-date).
I noticed that even when the bot is created for only one server, admins like when it creates his own role, as the other bots do, so they can directly customize permissions for each bot and drag them at the right place. (yes, that can be done with other roles, but they won't have the special status, making it assigned to the bot)

When I made this PR, the two main goals were making that command public and making it possible to edit the permissions integer, mostly what I had in mind. The `[p]inviteset` command is made to customize the invite link with all Discord's features, and the permissions interger is the main feature of an invite link, it would be weird to remove that...

Reply:
I'm not saying to remove the permissions integer, but to remove the redirect. Without extra work, this will just appear to "break" the invite link for most end users. Even if you do go through half of that work, without handling the remaining portions, you get things like inviting the bot, followed by being sent to a link like:

https://www.google.com/search?q=how+do+I+use+discord?code=sdfgbhbgvcxcvgbgvfcdx&guild_id=42424242&permissions=0

(not a real code or ID)

Beyond this, nothing in red is enabled to use this, so it would have to be coded themselves. 

Beyond this, you do not have to edit the pip package to edit commands included with the bot. We have paths to load from, and the ability to remove and replace commands.

I think this is asking for a headache telling users over and over that they need a site designed to use this just to say "hey, our command does everything an invite can", without considering if it should.

Reply:
Revisiting this: If you add a warning when attempting to add a redirect that this may break the invite if used improperly, it would be sufficient for "well, you didn't listen to the warning"